pkg/azuredisk/nodeserver_v2.go

```diff
@@ -91,25 +135,30 @@ func (d *DriverV2) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolume
 		return nil, status.Errorf(codes.Internal, "Failed to find disk on lun %s. %v", lunStr, err)
 	}
 
-	source, err := d.nodeProvisioner.GetDevicePathWithLUN(int(lun))
-	if err != nil {
-		// check if AzVolumeAttachment reports attached
-		if state, getErr := d.crdProvisioner.GetAzVolumeAttachmentState(ctx, diskURI, d.NodeID); state == nil || getErr != nil {
-			getErr = status.Errorf(codes.Internal, "failed to get current attachment state for volume (%s) and node (%s): %v", diskURI, d.NodeID, getErr)
-			klog.Error(getErr)
-			return nil, getErr
+	// check if AzVolumeattachment is in Attached state if recovery is in process for the specified volume and is not in process of detaching
+	if _, diskURIMatches := d.isRecoveryInProcess(diskURI, false); diskURIMatches {
```

Comment:
Passing a bool for acquiring or not acquiring a lock is very error prone and can result in a deadlock if someone accidentally passes a wrong value. `isRecoveryInProcess` should be divided into 2 functions since it is technically doing more than one thing. One `isRecoveryInProcess` that checks if `deviceChecker.entry` is not null without needing a lock and `doesDiskURIMatch` that always assumes an external lock and checks for uri equality.

Reply:
We need a lock to access the pointer itself. So both checks (1. whether `deviceChecker.entry == nil` and 2. whether `deviceChecker.entry.DiskURI == diskURI`) require a lock. And although `isRecoveryInProcess` is doing two things technically but given the usage of the function, I don't think we should divide the function into two not to mention taht the two checks are very trivial,.

And I agree that passing a boolean to note whether lock should be acquired or not can be error prone but the reasoning behind this is that for some functions where `isRecoveryInProcess` called (e.g. `shouldStartRecovery` and many more), we want the entire flow to be atomic. Otherwise, these functions can fail from a race condition.

Reply:
`getDetachState` already has a bug of acquiring the lock that is already acquired. If we do the a lock for all access types, we should assume the lock is already acquired for the function.

Reply:
We can document that this function should be strictly used with lock acquired and proceed without trying to acquire a lock within the function.